src/main/java/org/mule/extension/sftp/internal/operation/BaseFileSystemOperations.java

```diff
@@ -0,0 +1,360 @@
+   * @param timeBetweenSizeCheck wait time between size checks to determine if a file is ready to be read in milliseconds.
+   * @param subsetList           parameter group that lets you obtain a subset of the results
+   * @return a {@link PagingProvider} of {@link Result} objects each one containing each file's content in the payload and
+   *         metadata in the attributes
+   * @throws IllegalArgumentException if {@code directoryPath} points to a file which doesn't exist or is not a directory
+   */
+  protected PagingProvider<FileSystem, Result<CursorProvider, org.mule.extension.sftp.api.FileAttributes>> doPagedList(FileConnectorConfig config,
+                                                                                                                       String directoryPath,
+                                                                                                                       boolean recursive,
+                                                                                                                       FileMatcher matchWith,
+                                                                                                                       Long timeBetweenSizeCheck,
+                                                                                                                       StreamingHelper streamingHelper,
+                                                                                                                       SubsetList subsetList) {
+    return new PagingProvider<FileSystem, Result<CursorProvider, org.mule.extension.sftp.api.FileAttributes>>() {
+
+      private List<Result<InputStream, org.mule.extension.sftp.api.FileAttributes>> files;
+      private Iterator<Result<InputStream, org.mule.extension.sftp.api.FileAttributes>> filesIterator;
+      private final AtomicBoolean initialised = new AtomicBoolean(false);
+
+      @Override
+      public List<Result<CursorProvider, org.mule.extension.sftp.api.FileAttributes>> getPage(FileSystem connection) {
+        if (initialised.compareAndSet(false, true)) {
+          initializePagingProvider(connection);
+        }
+        List<Result<CursorProvider, org.mule.extension.sftp.api.FileAttributes>> page = new LinkedList<>();
+        for (int i = 0; i < LIST_PAGE_SIZE && filesIterator.hasNext(); i++) {
+          Result<InputStream, org.mule.extension.sftp.api.FileAttributes> result = filesIterator.next();
+          page.add((Result.<CursorProvider, org.mule.extension.sftp.api.FileAttributes>builder()
+              .attributes(result.getAttributes().get())
+              .output((CursorProvider) streamingHelper.resolveCursorProvider(result.getOutput()))
+              .mediaType(result.getMediaType().orElse(null))
+              .attributesMediaType(result.getAttributesMediaType().orElse(null))
+              .build()));
+        }
+        return page;
+      }
```

Comment:
This looks unnecessary. Why do you have a paging mechanism if the actual query against the service is not split into pages? See [SftpClient->list](https://github.com/mulesoft/mule-sftp-connector/blob/feature/W-12049455/src/main/java/org/mule/extension/sftp/internal/connection/SftpClient.java#L292)